### PR TITLE
Core/adding global pointer serialization

### PR DESCRIPTION
**Pull request by RiccardoRossi**
this branch adding a container for global pointers with a working serialization mechanism.

the save and load functions use specialized save and load operators of global pointer.

**a somewhat specific behaviour however is that the object to which the GlobalPointer points to is not serialized (only the address of it is)**

### CODE REVIEW

File: kratos/containers/global_pointers_vector.h [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+  
+  ///@} 
+  ///@name Input and output 
+  ///@{ 
+        
+ 
+  /// input stream function
+  template< class TDataType >
+  inline std::istream& operator >> (std::istream& rIStream, 
+				    GlobalPointersVector<TDataType>& rThis){}
```
Comment (roigcarlo) [inline kratos/containers/global_pointers_vector.h]:
```suggestion
				    GlobalPointersVector<TDataType>& rThis){ return rIStream;}
```

Comment (roigcarlo) [reply kratos/containers/global_pointers_vector.h]: That should fix the travis error

**Comment by pooyan-dadvand** [comment]
I see a technialc

**Comment by pooyan-dadvand** [comment]
Sorry I closed it accidentally

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
This implementation will break the use of global pointer as a local pointer after load....
Consider:
```c++
GlobalPointer p_node(0, &some_node); // points to position of some_node which is 0x1A1A in memory
save(p_node) //saving the 0x1A1A
save(some_node) 
....
load (p_node)   // now points to 0x1A1A 
                // which may be out of application memory: using it -> segmentation fault
load(some_node) // it is allocated in 0x1234 and p_node is not pointing to it!
```

**Comment by roigcarlo** [comment]
Agree with @pooyan-dadvand, but I think @RiccardoRossi code is legit even in the scenario you suggest.

GlobalPointer was not ment to be serializable in the first place (hence the lack of save/load functions) and as I told @RiccardoRossi I can't remember the exact reasons for that decision but it may perfectly be related to this.

Once you serialize your global pointer you need to decide how they will behave: 

### 1 - You follow the pointer chain
and serialize everything, you cannot use such serializer to transfer them over processes as they will lose their relevance:

```C++
// Process 1
auto Node::Pointer some_node; // some_node = 0x1000
GlobalPointer some_node_gp:   // some_node_gp(@) -> 0x1000, some_node_gp(proc) -> 1
    // We serialize following the pointer chain

// Process 2 
    // We deserialize
auto Node::Pointer some_node; // some_node = 0x2000
GlobalPointer some_node_gp:   // some_node_gp(@) -> 0x2000, some_node_gp(proc) -> 1
```

Which doesn't make any sense as the GlobalPointer should keep the pointer of the original proc. But if you just save to a file and restore in the same proc the result will be ok.

### 2 - You keep the pointer
```C++
// Process 1
auto Node::Pointer some_node; // some_node = 0x1000
GlobalPointer some_node_gp:   // some_node_gp(@) -> 0x1000, some_node_gp(proc) -> 1
    // We serialize following the pointer chain

// Process 2 
    // We deserialize
auto Node::Pointer some_node; // some_node = 0x2000
GlobalPointer some_node_gp:   // some_node_gp(@) -> 0x1000, some_node_gp(proc) -> 1
```

Which is ok from the point of view of the GlobalPointer functionality, but as @pooyan-dadvand  points out will not work if you are trying to restore a process: the pointer will be GlobalPointer.

The only solution I see for this is to decide which behavior we want or make a specialized version of the serializer which handles the pointer differently. In which case we should move the save/load function of the global pointer inside the serializer itself.

**Comment by RiccardoRossi** [comment]
From my point of view, the meaning of a GlobalPointer is closely related to that of a Weak Pointer.

I believe that the content of a GlobalPointer should NOT be serialized, that is, that the only relevant information to be serialized is the pointer address.
This however is one of the cases in which the meaning of serialization as "stop and restart" is very different from the use as mpi transfer.

From my point of view, the Stop-and-restart behaviour that we should have is a sort of "delayed loading".

With this i mean that in the load function we should construct  a list of the global pointers we load, pointing to the address that was originally stored.
At the end of the loading, when all the objecs are deserialized, we could go through this list and reassign the addresses according to the [saved_address, loaded_address] table constructed in the serializer.

this would mean that the global pointer containers would need to have two functions, one serializing to an array of addresses (without ant change) and a "real" save/load with the delayed mechanism. The second one could use internally the first, while the first would also have its use for efficient mpi transfer

**Comment by pooyan-dadvand** [comment]
I see....
Nevertheless, I'm pretty sure that if we put the save and load someone would store it...
Maybe we should add some flags to serializer to distinguish between MPI, restart, dump, memory count, etc...:
```c++
if(serializer.Is(Serializer::MPI))
    save(pointer_address);
else
    KRATOS_ERROR << "you may not use global pointer for save and load";
```
or 
```c++
if(serializer.Is(Serializer::MPI))
    save(pointer_address);
else if(serializer.Is(Serializer::RESTART))
    save(pointer);
// do nothing for memory count ....
```
About the weak_ptr I disagree, I think they are valid cases in which they have to be saved and loaded... (i.e. parent element in hierarchical elements)

**Comment by RiccardoRossi** [comment]
unfortunately it is not trivial to add flags to the serializer. The problem is that the serializer itself is used everywhere (and in particular in the flags) so including the flags in the serializer generates a cyclic include.

i could have a sort of behavioural flag included in the serializer...

**Review by RiccardoRossi — COMMENTED** [review]
@pooyan-dadvand serializer is now derived from flags as agreed. This required some gymnastics with a forward declaration of serializer in flags, but it doesn't look too bad

**Review by RiccardoRossi — COMMENTED** [review]
@pooyan-dadvand i think this is now ready for being in master

**Comment by RiccardoRossi** [comment]
approving anyway, since the epsilon i pointed to has a different meaning

[EDIT:] posted in the wrong PR , forget about this post :-D

**Review by pooyan-dadvand — APPROVED** [review]
I would have used `static_cast` instead of `interpreter_cast` as they don't change the pointer address for alignment. However in your case the alignment is the same and the behaviour should be the same.

**Comment by RiccardoRossi** [comment]
leaving this branch opened to try changing reinterprete_cast -> static_cast

**Comment by loumalouomega** [comment]
@RiccardoRossi do you still need this branch?

**Comment by RiccardoRossi** [comment]
no, please.feel free to remove, thx

**Pull request merged**